Song/track/handler for url refactoring attempt

philippe44 opened the pull request:
I decided to give it a try. There is fair bit of changes, some are more cosmetics, I realize, but they help making the separation between $song and $track more clear as it seems that the concept has evolved with LMS. For example, in Slim;;Player::Playlist.pm it was the same object at the beginning.

I've probably made more harm than good but well, the idea was that if I try something, I'll go all the way. Reading the evolution of LMS accross time, I think that distinction between $handler and $songHandler was necessary a long while ago (7.3) but there was a time when new firmware came that seems to have made that not required.

The idea is to use handlerFromUrl only when there is no $song and the $song->currentTrackHandler otherwise. If, after a scanUrl call, the track is modified and the currentTrackHandler is re-set because scanUrl has really changed the url (like unwrapped it to a different PH) then it will change as of before otherwise it will continue to call the PH set at creation of $song. With a modification I've made in the podcast PR, the PH has the possibility to keep control of the track if it wants to as LMS asks if it can change PH after the scanUrl. Now, if you decide to go further with that direction, the podcats PR will be need to be merged.

Now, this clearly reaches the limit of having no test harness and so I have no idea how much I've broken things.

philippe44 committed:
attempt to refactor handlerForUrl & song vs track naming

philippe44 committed:
fix open items

michaelherger commented on Slim/Player/Playlist.pm:
```diff
@@ -52,7 +52,12 @@ sub shuffleType {
 
-sub song {
+sub song { 
+	$log->warn("use of Slim::Player::Playlist::song is deprecated, use track instead");
```
You could add a `$log->is_info && logBacktrace('deprecated!')` here. We might want to know what was calling us the wrong way.

philippe44 replied:
Done

michaelherger reviewed (approved):
Wow... this looks almost too simple and clean to be true. I certainly missed a detail, but overall this looks like a very nice cleanup. I'd be happy to merge this to 8.2 and see what happens. Anything you wanted to add before I go ahead?

I should have another pass to see whether we're still using `handlerFromUrl` where we should no longer be (you didn't do this, did you?). Certainly for my plugins, too. But that can be done after a merge.

philippe44 committed:
move back to songHandler & urlHandler

They are both needed (surprise...) when PH's don't descend from HTTP. Now, the requestString needed to be updated

philippe44 committed:
use $song in getMetadataFor only for current song

philippe44 commented:
> Wow... this looks almost too simple and clean to be true. I certainly missed a detail, but overall this looks like a very nice cleanup. I'd be happy to merge this to 8.2 and see what happens. Anything you wanted to add before I go ahead?
> 
> I should have another pass to see whether we're still using `handlerFromUrl` where we should no longer be (you didn't do this, did you?). Certainly for my plugins, too. But that can be done after a merge.
>
Unfortunately, I tripped again and there was more than what I saw initially (surprise!). Both the songHandler and handler were needed because some methods need to exist even if the ancestor of the $song's PH is not a HTTP (or a RemoteStream). 

For example, if a custom PH wants to autorise direct streaming and for that returns a plain HTTP url in canDirectStream but this custom PH is not a subclass of HTTP/RemoteStream, then such requestString does not exist in that PH, although it exists in S:P::P::HTTP, which is the PH for the true HTTP direct stream url. Now, if the custom PH wants to redefine requestString even for direct streaming, it should be able to (see mixcloud plugin). So the general rule is that the song's protocol handler is tried first and if it does not have the methods, then the url's one is tried. 

At the end, this PR contains much less real changes than I was afraid of. I don't know if it's good or bad, but "la montagne a accouché d'une souris :smile:".  Most changes are really cosmetics and if you don't like them, we can remove these. I just thought it would help future developers as cosmetic is mainly for consistency (although there are a few cases where it was purely my preference...). For example, I've modified SongStreamController a lot, but there is nothing new. I was just surprised that it was not bases on an accessor and I could not see why it was manually doing the exact same thing an accessor does.

I also took the liberty to add a DEVELOPERS.txt file where I thought we could add some documentation on features we add that can be reused in plugins or can simply help future contributors.

Now, I think I need to continue looking at these changes as I feel there is more to review

philippe44 commented on Slim/Player/Protocols/HTTP.pm:
```diff
@@ -1105,11 +1107,11 @@ sub getMetadataFor {
 	}
 	else {
+		# make sure that protocol handler is what the $song wanted, not just the $url-based one
+		my $handler = $forPlaying ? $song->currentTrackHandler : Slim::Player::ProtocolHandlers->handlerForURL($url);
```
This is an important one if we want to make sure that the custom PH for the url track being played (the song) is called, even if the url is using a different PH

philippe44 commented on Slim/Player/SoftSqueeze.pm:
```diff
@@ -99,16 +99,15 @@ sub canDecodeRhapsody { 0 };
 sub hasScrolling { 0 }
 
 sub canDirectStream {
```
This is cosmetic

philippe44 commented on Slim/Player/Squeezebox.pm:
```diff
@@ -563,9 +563,8 @@ sub stream_s {
-		if ( !$format && $handler->can("getFormatForURL") ) {
-			$format = $handler->getFormatForURL($url);
-		}
+		my $methodHandler = $songHandler->can('getFormatForURL') ? $songHandler : $handler;
```
We need to give priority to $song's PH (that can always default to url's handler if needs to)

philippe44 commented on Slim/Player/Squeezebox.pm:
```diff
@@ -773,7 +772,8 @@ sub stream_s {
 
 		main::INFOLOG && logger('player.streaming.direct')->info("SqueezePlay direct stream: $url");
 
-		$request_string = $songHandler->requestString($client, $url, undef, $params->{'seekdata'});
```
If the custom PH allowed a directs stream and is not a subclass of HTTP, then it does not have a requestString method and that would crash if we only use songHandler. But, that PH might also want to overload requestString, so try it first

philippe44 commented on Slim/Player/Squeezebox2.pm:
```diff
@@ -458,19 +458,17 @@ sub songElapsedSeconds {
 }
 
 sub canDirectStream {
```
Mostly cosmetic

philippe44 commented on Slim/Player/SongStreamController.pm:
```diff
@@ -1,6 +1,5 @@
 package Slim::Player::SongStreamController;
```
Mostly cosmetic to make it inherit from accessor. Also, I've changed protocolHandler to urlHandler so that we have consistent naming between songHandler (for $song), streamHandler (for $sock) and urlHandler (for $song->streamUrl)

philippe44 committed:
undo cosmetic changes

philippe44 committed:
use songhandler core, not track

philippe44 committed:
fixes

philippe44 committed:
use currentTrack when possible

philippe44 committed:
oversight: streamUrl must be used in canDirectStreamSong

philippe44 committed:
a few more

philippe44 committed:
remove canDirectStreamSong in PH

This is un-necesseary  as the objective was to call SUPER::canDirectStream. Now that we have a S::P::P::HTTP(S)::canDirectStreamSong, it will take care of that

philippe44 commented:
ok, with the #611 PR, I think it's ready to see what will be user's feedback. Note that I've now changed a few PH to remove CanDirectStreamSong as it was just transferring control to ancestor's canDirectStream. Now that we have a canDirectStreamSong in their ancestors, it is no longer necessary but should help making things more clear for future PH developers, I hope

philippe44 committed:
adding some random comments

michaelherger commented on DEVELOPERS.txt:
```diff
@@ -0,0 +1,260 @@
+during the scan. Ultimately, when the actual image is cached (not the link), there will be
+an artwork cache entry that getMetadataFor() will use. 
+
+Note than some format may have other cover art inside streamed with the track and, for 
```
`Note thaT...`?

philippe44 replied:
done

michaelherger commented on DEVELOPERS.txt:
```diff
@@ -0,0 +1,260 @@
+------------- How is image/icon/cover passed from the OPML item to a track -------------
+
+When a $url from an OPML item is added to current playlist, a $track object is created 
+and its image is obtained by calling handlerForUrl($url)->getMetadataFor($url). 
+If this protocol handler is HTTP, it needs to find the image just by using this $url, 
+there is no other information. 
```
This part is not limited to OPML items, is it? That would apply to any track added/played, wouldn't it? The OPML special sauce is below mentioned calling of `setRemoteMetadata`.

philippe44 replied:
Probably, but this part of LMS is really a big unkown to me

michaelherger commented on DEVELOPERS.txt:
```diff
@@ -0,0 +1,260 @@
+from $args->{'song'}->track->url. When it finally returns, scanUrl provides a new $track
+that replaces the current one in the $song object and in the playlist.
+
+------------------------ Thin Protocol Hander (e.g. podcast) -----------------------------
```
`HandLer`

philippe44 replied:
done

michaelherger commented on DEVELOPERS.txt:
```diff
@@ -0,0 +1,260 @@
+Slim::Player::Protocol::HTTP uses that to replace itself by its HTTPS version when there is a
+HTTP->HTTPS upgrade. So the default behaviour is to *keep* the existing protocol handler.
+
+3- Now, not every thin protocol handler's methods will be called after this scan because some
```
Where's 2? 😁

philippe44 replied:
Lost in translation :smile:

michaelherger commented on DEVELOPERS.txt:
```diff
@@ -0,0 +1,260 @@
+1- When a $track object has changed after scanUrl, the required protocol handler might need to 
+be re-evaluated and replaced by what matches the actual $track->url (there are many protocol 
+handlers but that's too difficult to describe here). It should not be re-evaluated all the time 
+otehrwise the thin protocol handler would loose control if the $track->url has been reset to 
```
`otherwise`

philippe44 replied:
done

michaelherger commented on DEVELOPERS.txt:
```diff
@@ -0,0 +1,260 @@
+
+When resuming, LMS sometimes crates a new() song, sometimes it just re-opens it with open(). In
+both cases, the $url inside the track is the one that scanUrl has put after redirection. To solve 
+that, if Slim::Player::Song::open() fails and there track has been redirected, open() will recurse
```
`there track` -> `the track`

philippe44 replied:
done

michaelherger commented on DEVELOPERS.txt:
```diff
@@ -0,0 +1,260 @@
+always use the original url if any. There seem to be no reason/arm in new() to try with the 
+redirected url as it will do a scanUrl anyway. 
+
+Note that the original url is only set when Slim::Utils::Misc::Scanner!!Remote::scanURL if called, 
```
Not sure about the `!!` and `is called`?

philippe44 replied:
done

michaelherger commented on DEVELOPERS.txt:
```diff
@@ -0,0 +1,260 @@
+	
+--------------------------- Some comments on tracks/song data structure ---------------------------	
+
+>>> URL's in $song
```
`$song` is a variable name... we can't rely on a variable name to define what is in it. What would that be if not a `Slim::Player::Song`?

philippe44 replied:
yes

michaelherger reviewed (approved):
Let's get this done. I've added a few comments. But these can be addressed after the fact.

michaelherger commented on Slim/Player/Protocols/HTTP.pm:
```diff
@@ -1029,7 +1030,9 @@ sub getMetadataFor {
 	# use HTTP protocol handler
 	my $song = $client->playingSong();
-	if ( $song && $song->track->url eq $url ) {
+	my $current = $song->currentTrack->url eq $url if $song;
+
+	if ( $current ) {
 		if ( my $meta = $song->pluginData('wmaMeta') ) {
```
I have a plugin which uses `wmaMeta` to override the metadata shown. This no longer work, as `$current` is not true. Where shall I start the investigation?

philippe44 replied:
I've updated the PR with what I think is the solution: need to check for the track and currentTrack as well

michaelherger replied:
Which PR?

michaelherger replied:
Oh, the new one, of course.
